src/agent/coverage/src/cobertura.rs

```diff
@@ -1,12 +1,56 @@
 use crate::source::SourceCoverage;
-use crate::source::SourceCoverageLocation;
 use crate::source::SourceFileCoverage;
 use anyhow::Context;
 use anyhow::Error;
 use anyhow::Result;
+use std::path::Path;
 use std::time::{SystemTime, UNIX_EPOCH};
 use xml::writer::{EmitterConfig, XmlEvent};
 
+pub fn compute_line_rate(line_values: Vec<u32>) -> f32 {
+    let mut line_rate = 0_f32;
+    if line_values[1] > 0 {
+        line_rate = line_values[1] as f32 / line_values[0] as f32;
+    }
+    line_rate
+}
+
+pub fn compute_line_values_coverage(files: Vec<SourceFileCoverage>) -> (Vec<u32>, f32) {
+    let mut line_values: Vec<u32> = Vec::new();
+    let mut valid_lines = 0;
+    let mut hit_lines = 0;
+    for file in files {
+        let locations = file.locations;
+        for location in locations {
+            valid_lines += 1;
+            if location.count > 0 {
+                hit_lines += 1;
+            }
+        }
+    }
+    line_values.push(valid_lines);
+    line_values.push(hit_lines);
+    let line_rate = compute_line_rate(line_values.clone());
+    (line_values, line_rate)
```

Comment:
In Rust, we can create types to help us out. They can also make code a little more self-documenting.

When we read a type like `(Vec<u32>, f32)`, you have to read where it came from to know what it means.

On the other hand, suppose we defined a struct like:
```rust
struct LineCoverage {
    pub hit: u64,
    pub valid: u64,
    pub rate: f64,
}

impl LineCoverage {
    pub fn new(hit: u64, valid: u64) -> Self {
        let rate = if valid == 0 {
            0.0
        } else {
            (hit as f64) / (valid as f64)
        };

        Self { hit, valid, rate }
    }
}
```
Now we know what everything means just from , and the rate calculation is encapsulated in the type construction.

Reply:
Created new type to have line value information.